DataDog/libdatadog
# Add metrics package for dogstatsd metric parsing

# What does this PR do?

This PR adds a `Metrics` crate that introduces a metrics agent. This agent has the ability to listen on a UDP socket, receive {statsd/dogstatsd} payloads, and flush them to Datadog.

# Motivation

This PR is the first of a larger effort to get a micro dogstatsd server implementation in Rust, to be used by Serverless workloads. We want customers in Google Cloud Functions, Azure Functions, etc. to be able to send statsd and dogstatsd-based metrics to Datadog.

# Additional Notes

Currently, we don't support sampling, and **only** support statsd/dogstatsd distribution metrics.

# How to test the change?

The PR includes tests, however, you can spin up a tiny Rust binary:

```rust
use std::env;

use metrics::agent::{ConfigBuilder, MetricsAgent};

#[tokio::main]
async fn main() {
    let config = ConfigBuilder::new()
        .api_key(env::var("DD_API_KEY").unwrap())
        .build()
        .expect("Error parsing config");

    MetricsAgent::with_config(config).run().await;
}
```

And send UDP packets to the endpoint yourself:

```bash
echo -n "ivan.test.metric:1|d|#tag1:val1" | nc -4u -w0 localhost 8125
```

On an unrelated note, this is also my first PR in Rust! :-)
